File: core/src/kotlin/fr/mesabloo/heavymachdefense/log/ColoredLogger.kt
package fr.mesabloo.heavymachdefense.log

import com.badlogic.gdx.ApplicationLogger
import java.util.*

/**
 * A colored Gdx logger which logs to [System.err].
 *
 *  * Log messages are colored in Blue
 *  * Debug messages are colored in Magenta
 *  * Error messages are colored in Red
 *
 * Note: This uses ANSI color codes, which should now be supported on all major
 * modern OSes (Linux, macOS and Windows).
 * In case your terminal does not support ANSI color codes, the output should contain
 * junk characters (e.g. `☐[32m`).
 */
class ColoredLogger : ApplicationLogger {
    override fun log(name: String, msg: String) {
        System.err.println("\u001B[32m[${Date()} - $name -  LOG \u001B[0m] $msg")
    }

    override fun log(name: String, msg: String, throwable: Throwable) {
        System.err.println("\u001B[32m[${Date()} - $name -  LOG \u001B[0m] $msg\n$throwable")
    }

    override fun error(name: String, msg: String) {
        System.err.println("\u001B[31m[${Date()} - $name - ERROR\u001B[0m] $msg")
    }

    override fun error(name: String, msg: String, throwable: Throwable) {
        System.err.println("\u001B[31m[${Date()} - $name - ERROR\u001B[0m] $msg\n$throwable")
    }

    override fun debug(name: String, msg: String) {
        System.err.println("\u001B[36m[${Date()} - $name - DEBUG\u001B[0m] $msg")
    }

    override fun debug(name: String, msg: String, throwable: Throwable) {
        System.err.println("\u001B[36m[${Date()} - $name - DEBUG\u001B[0m] $msg\n$throwable")
    }
}

File: core/src/main/scala/io/github/mesabloo/hmdefense/logging/ColoredLogger.scala
package io.github.mesabloo.hmdefense.logging

import com.badlogic.gdx.ApplicationLogger

import java.util.Date
import scala.io.AnsiColor.*

/** A GDX logger outputting to [[Console.err]] with colors:
  *
  *   - Blue for logs;
  *   - Red for errors;
  *   - Magenta for debugs.
  */
object ColoredLogger extends ApplicationLogger:
  final override def log(tag: String, message: String): Unit =
    Console.err.println(s"$BOLD$BLUE${Date()} ─ $tag [i]$RESET $message")

  final override def log(
      tag: String,
      message: String,
      exception: Throwable
  ): Unit =
    this.log(tag, s"$message\n$exception")

  final override def error(tag: String, message: String): Unit =
    Console.err.println(s"$BOLD$RED${Date()} ─ $tag [!]$RESET $message")

  final override def error(
      tag: String,
      message: String,
      exception: Throwable
  ): Unit =
    this.log(tag, s"$message\n$exception")

  final override def debug(tag: String, message: String): Unit =
    Console.err.println(s"$BOLD$MAGENTA${Date()} ─ $tag [?]$RESET $message")

  final override def debug(
      tag: String,
      message: String,
      exception: Throwable
  ): Unit =
    this.debug(tag, s"$message\n$exception")
